### PR TITLE
bilinear upsampling and resizing

### PR DESCRIPTION
**Pull request by ahundt**
This is an implementation of bilinear interpolation for the backend and layers. It introduces the parameter `resize_images(interpolation='bilinear')` and `UpSampling2D(interpolation='bilinear')`.

All defaults remain `interpolation='nearest'` for backwards compatibility.

### CODE REVIEW

File: keras/backend/cntk_backend.py [hunk 1]
```diff
@@ -1122,17 +1122,20 @@ def permute_dimensions(x, pattern):
...
+        if data_format == 'channels_first':
+            output = repeat_elements(x, height_factor, axis=2)
+            output = repeat_elements(output, width_factor, axis=3)
+            return output
+        elif data_format == 'channels_last':
+            output = repeat_elements(x, height_factor, axis=1)
+            output = repeat_elements(output, width_factor, axis=2)
+            return output
+        else:
+            raise ValueError('CNTK Backend: Invalid data_format:', data_format)
```
Comment (elephantgrass7) [inline keras/backend/cntk_backend.py]: maybe tell the user the valid data_formats

File: keras/backend/tensorflow_backend.py [hunk 2]
```diff
@@ -1925,7 +1926,12 @@ def resize_images(x, height_factor, width_factor, data_format):
         new_shape = tf.shape(x)[2:]
         new_shape *= tf.constant(np.array([height_factor, width_factor]).astype('int32'))
         x = permute_dimensions(x, [0, 2, 3, 1])
-        x = tf.image.resize_nearest_neighbor(x, new_shape)
+        if interpolation == 'nearest':
+            x = tf.image.resize_nearest_neighbor(x, new_shape)
+        elif interpolation == 'bilinear':
+            x = tf.image.resize_bilinear(x, new_shape)
+        else:
```
Comment (elephantgrass7) [inline keras/backend/tensorflow_backend.py]: this value error is good but do in other places

File: keras/backend/theano_backend.py [hunk 3]
```diff
@@ -884,24 +884,42 @@ def repeat_elements(x, rep, axis):
...
 
-def resize_images(x, height_factor, width_factor, data_format):
+def resize_images(x, height_factor, width_factor, data_format, interpolation='nearest'):
     """Resize the images contained in a 4D tensor of shape
     - [batch, channels, height, width] (for 'channels_first' data_format)
     - [batch, height, width, channels] (for 'channels_last' data_format)
     by a factor of (height_factor, width_factor). Both factors should be
     positive integers.
     """
     if data_format == 'channels_first':
```
Comment (elephantgrass7) [inline keras/backend/theano_backend.py]: add comments to make easier to read

File: keras/layers/convolutional.py [hunk 4]
```diff
@@ -1565,6 +1565,7 @@ class UpSampling2D(Layer):
             It defaults to the `image_data_format` value found in your
             Keras config file at `~/.keras/keras.json`.
             If you never set it, then it will be "channels_last".
+        interpolation: A string, one of `nearest` or `bilinear`.
```
Comment (elephantgrass7) [inline keras/layers/convolutional.py]: other doc string stuff has default marked

File: keras/backend/cntk_backend.py (hunk 1 above)
Comment (elephantgrass7) [inline keras/backend/cntk_backend.py]: in the tests k.backend() == `cntk` might want to lower case here

File: tests/keras/layers/convolutional_test.py [hunk 5]
```diff
@@ -734,6 +734,43 @@ def test_upsampling_2d():
...
+                   input_shape=inputs.shape)
+
+        for length_row in [2]:
+            for length_col in [2, 3]:
+                layer = convolutional.UpSampling2D(
+                    size=(length_row, length_col),
+                    data_format=data_format)
+                layer.build(inputs.shape)
+                outputs = layer(K.variable(inputs))
+                np_output = K.eval(outputs)
```
Comment (elephantgrass7) [inline tests/keras/layers/convolutional_test.py]: this tests need to be more rigorous. only checking shape

File: keras/backend/cntk_backend.py (hunk 1 above)
Comment (brge17) [reply keras/backend/cntk_backend.py]: like the interpolation error in https://github.com/keras-team/keras/pull/9273/files

**Comment by ozabluda** [comment]
>This is an implementation of bilinear interpolation for the backend and layers. It introduces the parameter resize_images(interpolation='bilinear') and UpSampling2D(interpolation='bilinear').

>All defaults remain interpolation='nearest' for backwards compatibility.

default for `UpSampling2D` has to remain `nearest`, but maybe after PR https://github.com/keras-team/keras/pull/8849, default for `resize_images` should be `bilinear`?

BTW, this seems to be how you do it in CNTK:
https://microsoft.github.io/CNTK-R/reference/transform_scale.html

**Comment by ahundt** [comment]
Can anyone suggest how to fix the theano errors? I'm not really a theano user so I was just implementing my best guess.

File: keras/backend/theano_backend.py [hunk 6]
```diff
@@ -884,24 +884,42 @@ def repeat_elements(x, rep, axis):
...
         raise ValueError('Invalid data_format:', data_format)
 
+    if interpolation == 'nearest':
+        output = repeat_elements(x, height_factor, axis=axis_1)
+        output = repeat_elements(output, width_factor, axis=axis_2)
+    elif interpolation == 'bilinear':
+        ratio = height_factor / width_factor
+        th_padding = _preprocess_padding('same')
+        output = theano.tensor.nnet.abstract_conv.bilinear_upsampling(
+            x, ratio=ratio, border_mode=th_padding)
```
Comment (Dref360) [inline keras/backend/theano_backend.py]: `border_mode` is not required. The output is of value : (batch size, num_input_channels, input row size * row ratio,
        input column size * column ratio)

So fi the ratio is good, everything should be.

**Comment by bhack** [comment]
Any update?

**Comment by wtreible** [comment]
I'm new to this, but it seems pretty important that the resizing function should support not only bilinear interpolation, but also bicubic, at least when using the tf backend. Many of my colleagues who have tried Keras are surprised that it only supports nn interpolation, especially since upsampling is used in almost every hourglass-style network. 

Also, volume resizing w/ trilinear interp would help Keras keep up with Torch for 3D hourglass implementations.

**Comment by bhack** [comment]
@wtreible Exactly

**Comment by Dref360** [comment]
I don't think we should merge it till the Theano version is fixed. The documentation of Theano doesn't match what's happening.

It's a great feature, I just don't think leaving Theano behind TF/CNTK is a good idea.

**Comment by Dref360** [comment]
It's still being supported for at least 4 months. (https://groups.google.com/forum/#!topic/theano-users/7Poq8BZutbY)

My main concern is that it's the first *major* feature that Theano won't implement.

Currently, the only test Theano is not able to pass is: test_batchnorm_convnet_no_center_no_scale

Should we stop supporting Theano (ie. drop the CI) before October?
We could merge this PR and advise people to switch to another backend.

If we do so, this PR requires little changes (ie. throws if Theano) and the changes for TF are LGTM

**Comment by fchollet** [comment]
> Should we stop supporting Theano (ie. drop the CI) before October?

It's okay if new features we add in Keras aren't supported in Theano. Exiting workflows should keep working (and we should keep Theano on CI) for at least 1.5 years from now (Theano EoL + 1 year).

We will gradually drop features when Theano becomes a friction. Whenever it is "cheap" to make things work with Theano, we'll make it work. One recent example is that we dropped recurrent dropout support in Theano because the new implementation couldn't easily be made to work with Theano.

Basically: let's not make Theano a bottleneck of Keras development. But let's not break it when that can be avoided at a low cost.

This PR doesn't change the default behavior, so it would be okay to skip Theano support (won't break existing users).

File: keras/backend/theano_backend.py [hunk 7]
```diff
@@ -884,24 +884,42 @@ def repeat_elements(x, rep, axis):
...
+        axis_1 = 1
+        axis_2 = 2
     else:
         raise ValueError('Invalid data_format:', data_format)
 
+    if interpolation == 'nearest':
+        output = repeat_elements(x, height_factor, axis=axis_1)
+        output = repeat_elements(output, width_factor, axis=axis_2)
+    elif interpolation == 'bilinear':
+        ratio = height_factor / width_factor
```
Comment (joelthchao) [inline keras/backend/theano_backend.py]: For theano, `ratio` needs to be `integer`.
```python
ratio = height_factor // width_factor
```

File: keras/backend/theano_backend.py [hunk 8]
```diff
@@ -884,24 +884,42 @@ def repeat_elements(x, rep, axis):
...
+        th_padding = _preprocess_padding('same')
+        output = theano.tensor.nnet.abstract_conv.bilinear_upsampling(
+            x, ratio=ratio)
+        if hasattr(x, '_keras_shape'):
+            output._keras_shape = list(x._keras_shape)
+            repeat_dim_1 = x._keras_shape[axis_1]
+            repeat_dim_2 = x._keras_shape[axis_2]
+            output._keras_shape[axis_1] = repeat_dim_1 * rep
+            output._keras_shape[axis_2] = repeat_dim_2 * rep
```
Comment (joelthchao) [inline keras/backend/theano_backend.py]: Unresolved reference`rep`

**Comment by gabrieldemarmiesse** [comment]
I tried to take a stab at it. See #10994.

File: keras/backend/theano_backend.py (hunk 7 above)
Comment (ahundt) [reply keras/backend/theano_backend.py]: how should this be done? Am I correct that the values might be tensors and not something that can be cast to `int` on every backend?

File: keras/backend/theano_backend.py (hunk 8 above)
Comment (ahundt) [reply keras/backend/theano_backend.py]: hmm... it has been a while since I looked at this... what was rep supposed to be?

**Comment by ahundt** [comment]
oh cool just saw https://github.com/keras-team/keras/pull/10994, thanks @gabrieldemarmiesse !

**Comment by gabrieldemarmiesse** [comment]
@ahundt can you close the PR now that #10994  is merged? 

And thanks again for your work! I didn't have to change it much.
